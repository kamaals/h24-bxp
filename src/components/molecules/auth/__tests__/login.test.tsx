import {
  describe,
  it,
  expect,
  vi,
  afterEach,
  beforeEach,
  type Mock,
} from "vitest";
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import Login from "../login";
import { useRouter } from "next/navigation";
import userEvent from "@testing-library/user-event";

vi.mock("next/navigation", () => ({
  useRouter: vi.fn(),
}));

vi.mock("sonner", () => ({
  toast: {
    error: vi.fn(),
  },
}));

vi.mock("@/lib/client", () => ({
  signIn: {
    email: vi.fn(),
  },
}));

describe("Login", async () => {
  const pushMock = vi.fn();
  beforeEach(() => {
    cleanup();
    // @ts-expect-error: mocking the push function
    (useRouter as Mock<typeof useRouter>).mockReturnValue({
      push: pushMock,
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("Should match with snapshot", () => {
    const result = render(<Login />);
    expect(result).toMatchSnapshot();
  });

  it("Should trigger login", () => {
    render(<Login />);
    const button = screen.getAllByTestId("loading-button");
    button[0].click();
  });

  it("Should updates input values when user types", async () => {
    const user = userEvent.setup();
    render(<Login />);

    const emailInput = screen.getByTestId("email") as HTMLInputElement;
    const passwordInput = screen.getByTestId("password") as HTMLInputElement;

    await user.clear(emailInput);
    await user.clear(passwordInput);

    await user.type(emailInput, "john.doe@example.com");
    await user.type(passwordInput, "password123password123");

    expect(emailInput.value).toBe("john.doe@example.com");
    expect(passwordInput.value).toBe("password123password123");
  });

  it("Should trigger login", async () => {
    const user = userEvent.setup();
    render(<Login />);

    const emailInput = screen.getByTestId("email") as HTMLInputElement;
    const passwordInput = screen.getByTestId("password") as HTMLInputElement;
    const loginBtn = screen.getByTestId("loading-button") as HTMLInputElement;

    await user.type(emailInput, "john.doe@example.com");
    await user.type(passwordInput, "password123password123");

    await user.click(loginBtn);
    waitFor(() => {
      const loader = screen.getByTestId("loader") as HTMLInputElement;
      expect(loader).toBeInTheDocument();
    });
  });
});
