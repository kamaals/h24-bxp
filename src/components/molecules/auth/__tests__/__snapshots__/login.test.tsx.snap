// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 max-w-sm w-full relative overflow-hidden shadow-md mt-8"
        data-slot="card"
      >
        <div
          class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
          data-slot="card-header"
        >
          <div
            class="font-semibold text-lg md:text-xl"
            data-slot="card-title"
          >
            <h4>
              Sign In
            </h4>
          </div>
          <div
            class="text-muted-foreground text-xs md:text-sm"
            data-slot="card-description"
          >
            Enter your email below to login to your account
          </div>
        </div>
        <div
          class="px-6"
          data-slot="card-content"
        >
          <form>
            <div
              class="grid gap-4"
            >
              <div
                class="grid gap-2"
              >
                <div
                  class="grid gap-2"
                  data-slot="form-item"
                >
                  <div>
                    <label
                      class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
                      data-error="false"
                      data-slot="form-label"
                      for="«r0»-form-item"
                    >
                      Email
                    </label>
                    <p
                      class="text-muted-foreground text-xs"
                      data-slot="form-description"
                      id="«r0»-form-item-description"
                      style="margin-top: -.125rem;"
                    />
                  </div>
                  <div
                    aria-describedby="«r0»-form-item-description"
                    aria-invalid="false"
                    class="space-y-1"
                    data-slot="form-control"
                    id="«r0»-form-item"
                  >
                    <input
                      class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
                      data-slot="input"
                      data-testid="email"
                      name="email"
                      value="john@doe.aes"
                    />
                    <span
                      class="inline-block h-2 bg-emerald-300 w-2 ml-1 rounded-full"
                    />
                  </div>
                </div>
              </div>
              <div
                class="grid gap-2"
              >
                <div
                  class="grid gap-2"
                  data-slot="form-item"
                >
                  <div>
                    <label
                      class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
                      data-error="false"
                      data-slot="form-label"
                      for="«r1»-form-item"
                    >
                      Password
                    </label>
                    <p
                      class="text-muted-foreground text-xs"
                      data-slot="form-description"
                      id="«r1»-form-item-description"
                      style="margin-top: -.125rem;"
                    />
                  </div>
                  <div
                    aria-describedby="«r1»-form-item-description"
                    aria-invalid="false"
                    class="space-y-1"
                    data-slot="form-control"
                    id="«r1»-form-item"
                  >
                    <input
                      class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
                      data-slot="input"
                      data-testid="password"
                      name="password"
                      type="password"
                      value="Element@AI"
                    />
                    <span
                      class="inline-block h-2 bg-emerald-300 w-2 ml-1 rounded-full"
                    />
                  </div>
                </div>
              </div>
              <div
                class="py-4 flex justify-end animate"
              >
                <button
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 bg-gradient-to-r from-purple-400 via-pink-500 to-red-500 text-white shadow-3xl"
                  data-slot="button"
                  data-testid="loading-button"
                  type="submit"
                >
                  <span
                    class="font-bold"
                  >
                    Sign in
                  </span>
                </button>
              </div>
            </div>
          </form>
        </div>
        <div
          class="pointer-events-none absolute inset-0 rounded-[inherit] border border-transparent [mask-clip:padding-box,border-box] [mask-composite:intersect] [mask-image:linear-gradient(transparent,transparent),linear-gradient(#000,#000)]"
        >
          <div
            class="absolute aspect-square bg-gradient-to-l from-[var(--color-from)] via-[var(--color-to)] to-transparent"
            style="width: 50px; offset-path: rect(0 auto auto 0 round 50px); --color-from: #ffaa40; --color-to: #9c40ff; offset-distance: 0%;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 max-w-sm w-full relative overflow-hidden shadow-md mt-8"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
        data-slot="card-header"
      >
        <div
          class="font-semibold text-lg md:text-xl"
          data-slot="card-title"
        >
          <h4>
            Sign In
          </h4>
        </div>
        <div
          class="text-muted-foreground text-xs md:text-sm"
          data-slot="card-description"
        >
          Enter your email below to login to your account
        </div>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <form>
          <div
            class="grid gap-4"
          >
            <div
              class="grid gap-2"
            >
              <div
                class="grid gap-2"
                data-slot="form-item"
              >
                <div>
                  <label
                    class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
                    data-error="false"
                    data-slot="form-label"
                    for="«r0»-form-item"
                  >
                    Email
                  </label>
                  <p
                    class="text-muted-foreground text-xs"
                    data-slot="form-description"
                    id="«r0»-form-item-description"
                    style="margin-top: -.125rem;"
                  />
                </div>
                <div
                  aria-describedby="«r0»-form-item-description"
                  aria-invalid="false"
                  class="space-y-1"
                  data-slot="form-control"
                  id="«r0»-form-item"
                >
                  <input
                    class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
                    data-slot="input"
                    data-testid="email"
                    name="email"
                    value="john@doe.aes"
                  />
                  <span
                    class="inline-block h-2 bg-emerald-300 w-2 ml-1 rounded-full"
                  />
                </div>
              </div>
            </div>
            <div
              class="grid gap-2"
            >
              <div
                class="grid gap-2"
                data-slot="form-item"
              >
                <div>
                  <label
                    class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
                    data-error="false"
                    data-slot="form-label"
                    for="«r1»-form-item"
                  >
                    Password
                  </label>
                  <p
                    class="text-muted-foreground text-xs"
                    data-slot="form-description"
                    id="«r1»-form-item-description"
                    style="margin-top: -.125rem;"
                  />
                </div>
                <div
                  aria-describedby="«r1»-form-item-description"
                  aria-invalid="false"
                  class="space-y-1"
                  data-slot="form-control"
                  id="«r1»-form-item"
                >
                  <input
                    class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
                    data-slot="input"
                    data-testid="password"
                    name="password"
                    type="password"
                    value="Element@AI"
                  />
                  <span
                    class="inline-block h-2 bg-emerald-300 w-2 ml-1 rounded-full"
                  />
                </div>
              </div>
            </div>
            <div
              class="py-4 flex justify-end animate"
            >
              <button
                class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 bg-gradient-to-r from-purple-400 via-pink-500 to-red-500 text-white shadow-3xl"
                data-slot="button"
                data-testid="loading-button"
                type="submit"
              >
                <span
                  class="font-bold"
                >
                  Sign in
                </span>
              </button>
            </div>
          </div>
        </form>
      </div>
      <div
        class="pointer-events-none absolute inset-0 rounded-[inherit] border border-transparent [mask-clip:padding-box,border-box] [mask-composite:intersect] [mask-image:linear-gradient(transparent,transparent),linear-gradient(#000,#000)]"
      >
        <div
          class="absolute aspect-square bg-gradient-to-l from-[var(--color-from)] via-[var(--color-to)] to-transparent"
          style="width: 50px; offset-path: rect(0 auto auto 0 round 50px); --color-from: #ffaa40; --color-to: #9c40ff; offset-distance: 0%;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
